### PR TITLE
add ModelBroadCast to reduce the model clone time

### PR DESCRIPTION
**Pull request by SeaOfOcean**

### CODE REVIEW

**Comment by SeaOfOcean** [comment]
pr validate pass & test imageclassifier example ok

File: dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadCast.scala [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+import com.intel.analytics.bigdl.Module
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.tensor.{Storage, Tensor}
+import org.apache.spark.SparkContext
+import org.apache.spark.broadcast.Broadcast
+
+import scala.reflect.ClassTag
+
+class ModelBroadCast[T: ClassTag](implicit ev: TensorNumeric[T]) extends Serializable {
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadCast.scala]: change to modelBroadcast

File: dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadCast.scala [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  def cloneModel(): Module[T] = {
+    val localModel = broadcastModel.value.cloneModule()
+    putWeightBias(broadcastParameters.value, localModel)
+    localModel
+  }
+
+
+  private def getWeightBias(parameters: (Array[Tensor[T]], Array[Tensor[T]]))
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadCast.scala]: getAndClearWeightBias

File: dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadCast.scala [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  : Array[Tensor[T]] = {
+    var i = 0
+    val weightsBias = new Array[Tensor[T]](parameters._1.length)
+    while (i < parameters._1.length) {
+      if (parameters._1(i) != null) {
+        val wb = parameters._1(i)
+        weightsBias(i) = Tensor[T](Storage(wb.storage().array()),
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadCast.scala]: =Tensor[T](wb)

File: dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadCast.scala [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    localModel
+  }
+
+
+  private def getWeightBias(parameters: (Array[Tensor[T]], Array[Tensor[T]]))
+  : Array[Tensor[T]] = {
+    var i = 0
+    val weightsBias = new Array[Tensor[T]](parameters._1.length)
+    while (i < parameters._1.length) {
+      if (parameters._1(i) != null) {
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadCast.scala]: what if "parameters._1(i) == null"?

File: dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadCast.scala [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+  def broadcast(sc: SparkContext, model: Module[T]): this.type = {
+    val bcModel = model.cloneModule()
+    val weightsBias = getWeightBias(bcModel.parameters())
+    broadcastModel = sc.broadcast(bcModel)
+    broadcastParameters = sc.broadcast(weightsBias)
+    this
+  }
+
+  def cloneModel(): Module[T] = {
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadCast.scala]: change to value()

**Comment by jason-dai** [comment]
Please add some new unit tests for this

File: dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadCast.scala (hunk 4 above)
Comment (SeaOfOcean) [inline dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadCast.scala]: the initialized value is null when creating the array, so we don't need to set it to null

File: dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadCast.scala (hunk 3 above)
Comment (SeaOfOcean) [inline dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadCast.scala]: if I use weightsBias(i) = Tensor[T](wb)
followed by wb.set()
storage of weightsBias(i) is also set.

**Comment by SeaOfOcean** [comment]
PR-Validate pass

**Comment by jason-dai** [comment]
Please add more scala doc and comments; otherwise it looks good to me.

**Comment by SeaOfOcean** [comment]
add scala doc and comments

**Pull request merged**